packages/app-desktop/gui/ResizableLayout/utils/scaleLayoutItemSizes.ts

```diff
@@ -0,0 +1,123 @@
+import { produce } from 'immer';
+import iterateItems from './iterateItems';
+import { LayoutItem, LayoutItemDirection, Size } from './types';
+import { itemMinHeight, itemMinWidth } from './useLayoutItemSizes';
+import validateLayout from './validateLayout';
+import { cleanupAutoSizeContext, currentAutoSizeContext, ensureAutoSizeContext } from './autoSizeUtils';
+
+function clamp(value: number, minimum: number) {
+	return Math.max(minimum, Math.round(value));
+}
+
+function safeRatio(value: number) {
+	if (!Number.isFinite(value) || value <= 0) return 1;
+	return value;
+}
+
+// Scales panel sizes proportionally when the window is resized.
+// Tracks "natural" sizes (before min-size constraints) to avoid cumulative scaling errors
+// when the window is repeatedly resized to/from minimum dimensions.
+export default function scaleLayoutItemSizes(layout: LayoutItem, newRootSize: Size, widthRatio: number, heightRatio: number): LayoutItem {
+	const ratioX = safeRatio(widthRatio);
+	const ratioY = safeRatio(heightRatio);
+
+	// Previous window size calculated from ratio, used for fallback
+	const referenceRootWidth = ratioX ? newRootSize.width / ratioX : undefined;
+	const referenceRootHeight = ratioY ? newRootSize.height / ratioY : undefined;
+	const savedRootWidth = layout.context?.savedRootSize?.width;
+	const savedRootHeight = layout.context?.savedRootSize?.height;
+	const previousRootWidth = layout.width;
+	const previousRootHeight = layout.height;
+
+	const scaledLayout = produce(layout, (draft: LayoutItem) => {
+		iterateItems(draft, (_itemIndex: number, item: LayoutItem, parent: LayoutItem) => {
+			if (!parent) {
+				item.width = newRootSize.width;
+				item.height = newRootSize.height;
+				return true;
+			}
+
+			const autoSize = currentAutoSizeContext(item);
+
+			if (parent.direction === LayoutItemDirection.Row) {
+				const hasExplicitWidth = typeof item.width === 'number';
+				const hasNaturalWidth = typeof autoSize?.naturalWidth === 'number';
+
+				if (hasExplicitWidth || hasNaturalWidth) {
+					const ensuredAutoSize = autoSize ?? ensureAutoSizeContext(item);
+
+					// Store reference window width for this panel's natural width
+					if (typeof ensuredAutoSize.rootWidth !== 'number') {
+						const fallbackRootWidth = referenceRootWidth ?? savedRootWidth ?? previousRootWidth ?? newRootSize.width;
+						if (typeof fallbackRootWidth === 'number') ensuredAutoSize.rootWidth = fallbackRootWidth;
+					}
+
+					// Calculate natural width (before min-size clamping) if not already tracked
+					if (typeof ensuredAutoSize.naturalWidth !== 'number' && hasExplicitWidth) {
+						const targetReferenceWidth = ensuredAutoSize.rootWidth ?? referenceRootWidth ?? savedRootWidth ?? previousRootWidth ?? newRootSize.width;
```

Comment:
```suggestion
						const targetReferenceWidth = ensuredAutoSize.rootWidth ?? referenceRootWidth;
```

`referenceRootWidth` is never `undefined` (see comment above).